semantics/src/main/kotlin/com/strumenta/kolasu/semantics/linking/ReferenceResolverCache.kt

```diff
@@ -0,0 +1,41 @@
+package com.strumenta.kolasu.semantics.linking
+
+import com.strumenta.kolasu.model.ReferenceByName
+import kotlin.reflect.KProperty1
+
+/**
+ * Abstract storage for resolution results.
+ **/
+interface ReferenceResolverCache {
```

Comment:
Why do we need this? Can't we just change the references to store the new state?

Reply:
The reference resolver returns back the result of resolving a reference, which is then used by the SymbolResolver to update references. The reference resolver can be used just to retrieve the target of a reference, regardless of it being local/global resolved/not resolved. For example, I could use it in a type computer to retrieve the target of a reference I want to compute the type for.

Reply:
I am not sure I understand, what would be a use case for emptying this cache? This is used to avoid repeating computations during reference resolution, much like memoization was used before. Having the possibility to provide different implementations is only a plus for cases where we might want different solutions than in-memory maps.

In a similar way, a custom implementation of a resolution cache could handle allocation/deallocation logic accordingly. No policy is currently defined.

Reply:
I understand that the goal is not to perform resolution more than once.

What I do not understand is under which circumstances you could not just use the ReferenceByName itself as the cache: when asked to resolve a reference, if resolved is already set you just get the result from there. In case the resolved field is set to null, then yes, you would end up recalculating the reference (to find out it is null) each time. Is it to avoid this case that you use the cache? If so:
- I would explicitly explain this in the documentation
- I will open an issue about reflecting on how to distinguish between the states "Resolution not yet calculated" and "Resolution calculated and resulting in no match" in Kolasu 1.6